File: algorithms/[202]快乐数/solution.py
# topics = ["哈希表"]

from typing import Set


class Solution:
    def isHappy(self, n: int) -> bool:
        history: Set[int] = set()

        while n not in history:
            history.add(n)
            m = 0
            while n:
                m += (n % 10) ** 2
                n = n // 10

            if m == 1:
                return True
            else:
                n = m

        return False

File: algorithms/[202]快乐数/solution.ts
// topics = ["双指针"]

function isHappy(n: number): boolean {
  // 同 141.环形链表
  let slow = n;
  let fast = nextNum(n);
  while (fast != 1 && fast != slow) {
    slow = nextNum(slow);
    fast = nextNum(nextNum(fast));
  }
  return fast == 1;
}

function nextNum(num: number): number {
  let sum = 0;
  while (num > 0) {
    const temp = num % 10;
    sum += temp * temp;
    num = Math.floor(num / 10);
  }
  return sum;
}
